lib/Canonicalize/SimplifySwitch/SimplifySwitchPass.cpp

```diff
@@ -0,0 +1,234 @@
+  } else {
+    revng_abort();
+  }
+}
+
+static llvm::BasicBlock *getBlockFor(llvm::SwitchInst *Switch,
+                                     llvm::Constant *C) {
+  uint64_t TheConstant = cast<llvm::ConstantInt>(C)->getValue().getZExtValue();
+  for (const auto &I : Switch->cases()) {
+    const llvm::ConstantInt *CaseVal = I.getCaseValue();
+    uint64_t Case = CaseVal->getValue().getZExtValue();
+    if (Case == TheConstant)
+      return I.getCaseSuccessor();
+  }
+
+  return nullptr;
+}
+
+static bool runSimplifySwitch(llvm::Function &F,
+                              llvm::LazyValueInfo &LVI,
+                              llvm::DominatorTree &DT,
+                              DataMemoryOracle &MO) {
+  if (F.isDeclaration() || !F.hasExactDefinition())
+    return false;
+
+  for (auto &BB : F) {
+    for (auto &I : BB) {
+      auto Switch = dyn_cast<llvm::SwitchInst>(&I);
+      if (not Switch)
+        continue;
+
+      auto *Condition = Switch->getCondition();
+      DataFlowGraph::Limits Limits(1000 /*MaxPhiLike*/, 1 /*MaxLoad*/);
+      auto Results = ValueMaterializer::getValuesFor(Switch,
+                                                     Condition,
+                                                     MO,
+                                                     LVI,
+                                                     DT,
+                                                     Limits,
+                                                     Oracle::AdvancedValueInfo);
+      auto &DFG = Results.dataFlowGraph();
+      // Identify the only node in the data flow graph that's associated to an
+      // Oracle and use it as a switch condition.
+      bool BailOut = false;
+      BidirectionalNode<DataFlowNode> *StartNode = nullptr;
+
+      for (auto BidirectNode : DFG.nodes()) {
+        const DataFlowNode &Node = *BidirectNode;
+        // Ignore nodes that are constants in the DFG.
+        if (llvm::isa<llvm::ConstantInt>(Node.Value))
+          continue;
+
+        if (Node.UseOracle) {
+          if (StartNode != nullptr) {
+            BailOut = true;
+            break;
+          } else {
+            // TODO: Fix this const_cast.
+            auto BiNode = const_cast<BiDFNode *>(BidirectNode);
+            StartNode = BiNode;
+          }
+        }
+      }
+
+      if (StartNode == nullptr or BailOut)
+        continue;
+
+      if (not Results.values())
+        continue;
+
+      auto StartNodeInst = dyn_cast<llvm::Instruction>(StartNode->Value);
+      if (StartNodeInst and isa<llvm::PHINode>(StartNodeInst))
+        continue;
+      auto ConditionInst = dyn_cast<llvm::Instruction>(Condition);
+      if (ConditionInst and isa<llvm::PHINode>(ConditionInst))
+        continue;
+
+      if (StartNode->Value == Condition)
+        continue;
+
+      if (Results.values()->size() != Switch->getNumCases()) {
+        revng_log(SwitchOptLogger, "Unexpected values!");
+        continue;
+      }
```

Comment:
All these early continues are very mysterious.
Thoroughly comment all of them.

Also, why are we skipping `PHINode`s?

Reply:
Most of them are removed now.

We were skipping phis as a workaround, but it is not needed anymore.